### PR TITLE
add basic imagenette demo

### PR DESCRIPTION
**Pull request by brettkoonce**

### CODE REVIEW

File: Imagenette/data.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import TensorFlow
+import Foundation
```
Comment (brettkoonce) [inline Imagenette/data.swift]: I had to use the prebuilt nightlies to get this working on linux.  When I build from master adding the foundation flag like so:
    utils/build-script --foundation --enable-tensorflow-gpu --release-debuginfo
I get:
    data.swift:2:8: error: no such module 'Foundation'
What am I doing wrong/is there some obvious flag I'm missing?

File: Imagenette/main.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+let optimizer = SGD(for: model, learningRate: 0.001, momentum: 0.9, nesterov: true, scalarType: Float.self)
+
+print("Starting training...")
+for epoch in 1...20 {
+    var trainingLossSum: Float = 0
+    var trainingBatchCount = 0
+    let shuffledDataset = trainingImageDataset.combinedDataset.shuffled(sampleCount: Int64(totalTrainingImages), randomSeed: Int64(epoch))
+
+    for batch in shuffledDataset.batched(Int64(42)) {
+        let (labels, images) = (batch.label, batch.data)
```
Comment (brettkoonce) [inline Imagenette/main.swift]: Then I would like to do some data augmentation here using Raw.randomCrop, is there an example of how to do so?

File: Imagenette/data.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.imageLabels = Tensor<Int32>(newImageLabels)
+
+        self.combinedDataset =  Dataset(elements: Example(label: Tensor<Int32>(self.imageLabels), data: self.imageData))
+    }
+}
+
+func loadImageUsingTensorFlow(from fileURL: URL, size: (Int, Int)) -> [Float]? {
+    let loadedFile = Raw.readFile(filename: StringTensor(fileURL.absoluteString))
+    let loadedJpeg = Raw.decodeJpeg(contents: loadedFile, channels: 3, dctMethod: "")
+    let resizedImage = Raw.resizeBilinear(images: Tensor<UInt8>([loadedJpeg]), size: Tensor<Int32>([Int32(size.0), Int32(size.1)]))
```
Comment (brettkoonce) [inline Imagenette/data.swift]: and then here for the data augmentation I need the results of resizeBilinear (eg. Tensor<Float>) but not the resize, what is the best wya to do the cast?

**Comment by brettkoonce** [comment]
ps thx in advance 🙇‍♂️

File: Imagenette/data.swift (hunk 1 above)
Comment (rxwei) [reply Imagenette/data.swift]: You'll need a full toolchain build. Use `utils/build-toolchain-tensorflow`.

File: Imagenette/main.swift (hunk 2 above)
Comment (rxwei) [reply Imagenette/main.swift]: `tf.image.random_corp` is not a core op defined in C++, so it is not a raw op. It is implemented [here](https://github.com/tensorflow/tensorflow/blob/6612da89516247503f03ef76e974b51a434fb52e/tensorflow/python/ops/random_ops.py#L288) in Python. You can implement your own locally.

File: Imagenette/data.swift (hunk 3 above)
Comment (rxwei) [reply Imagenette/data.swift]: Not sure I understand what you mean by "not the resize". What exactly do you want to cast?

File: Imagenette/main.swift (hunk 2 above)
Comment (brettkoonce) [reply Imagenette/main.swift]: can I not use this?

(swift) Raw.randomCrop()
<REPL Input>:1:16: error: missing argument for parameter 'image' in call
Raw.randomCrop()
               ^
               image: <#Tensor<T>#>
TensorFlow.Raw:644:35: note: 'randomCrop(image:size:seed:seed2:)' declared here
    @inlinable public static func randomCrop<T>(image: TensorFlow.Tensor<T>, size: TensorFlow.Tensor<Int64>, seed: Int64 = 0, seed2: Int64 = 0) -> TensorFlow.Tensor<T> where T : Numeric, T : TensorFlowScalar

File: Imagenette/data.swift (hunk 3 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: This step resizes the image and converts it from uint8 to float, which the .scalars call after requires.  I am just trying to pass the image through as a float and then will randomCrop it later to get the correct input size, does that make sense?

File: Imagenette/data.swift (hunk 1 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: thanks, building away!

File: Imagenette/main.swift (hunk 2 above)
Comment (rxwei) [reply Imagenette/main.swift]: Oh, you are right it exists. Just use that then :)  We don't have an example of how to use that, however.

File: Imagenette/data.swift (hunk 3 above)
Comment (rxwei) [reply Imagenette/data.swift]: The [`.scalars`](https://github.com/apple/swift/blob/19c8d02990ac650b7bc0a4f69de1bbca0306a4bc/stdlib/public/TensorFlow/Tensor.swift#L871) property does not require the scalar type to be `Float`. It can be any scalar type.

File: Imagenette/main.swift (hunk 2 above)
Comment (brettkoonce) [reply Imagenette/main.swift]: you are correct 😅

`Fatal error: Op RandomCrop is not available in GraphDef version 29. It has been removed in version 8. Random crop is now pure Python.: file /swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2090
`

File: Imagenette/data.swift (hunk 3 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: think i figured out what's going on, thanks!

**Comment by brettkoonce** [comment]
@rxwei this is much cleaner and training properly, but a bit memory/cycle hungry.  i built a variant of this for real-time data augmentation (doing the crop+resize inside the training loop across a full batch) yesterday with cifar, where the data is much cleaner.

for imagenette, though, i need to way to store a collection of different sized images in a single tensor (eg [total_images, random_width, random_height, channels]) to pass around, is there a good way to deal with a collection of different images?  once i have said tensor I should be able to parse it in place/during the training loop.

File: Imagenette/data.swift (hunk 3 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: think i got it working, please see updated code!

File: Imagenette/main.swift (hunk 2 above)
Comment (brettkoonce) [reply Imagenette/main.swift]: see below for use of Raw.cropAndResize + manually computed random offsets

File: Imagenette/data.swift (hunk 1 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: it worked! 🚀

**Comment by brettkoonce** [comment]
@rxwei reworked the data pipeline, could use some refactoring, but it works!  thanks for the notes!

File: Imagenette/main.swift [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+let classCount = trainingImageDataset.classes
+let totalTrainingImages = trainingImageDataset.imageData.shape[0]
+let totalValidationImages = validationImageDataset.imageData.shape[0]
+
+var model = BasicCNNModel()
+let optimizer = SGD(for: model, learningRate: 0.002, momentum: 0.9, nesterov: true, scalarType: Float.self)
+
+print("Starting training...")
+for epoch in 1...80 {
+    if (epoch > 8) { optimizer.learningRate = 0.001 }
```
Comment (rxwei) [inline Imagenette/main.swift]: Swift conditionals do not need parentheses around the condition. Remove parentheses here. Same for other occurrences.

File: Imagenette/data.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    let loadedFile = Raw.readFile(filename: StringTensor(fileURL.absoluteString))
+    let loadedJpeg = Raw.decodeJpeg(contents: loadedFile, channels: 3, dctMethod: "")
+
+    let maxX: Float = Float(loadedJpeg.shape[0])
+    let maxY: Float = Float(loadedJpeg.shape[1])
+
+    let xPrime: Float = (Float(maxX) - Float(size.0))/2.0
+    let yPrime: Float = (Float(maxY) - Float(size.1))/2.0
+
+    let xOne = (xPrime)/maxX
```
Comment (rxwei) [inline Imagenette/data.swift]: Why parenthesize a single variable?
```suggestion
    let xOne = xPrime / maxX
```

File: Imagenette/data.swift [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    let yOne = (yPrime)/maxY
+    let xTwo = (xPrime + Float(size.0))/maxX
+    let yTwo = (yPrime + Float(size.1))/maxY
+
+    let boxesWrapped = Tensor<Float>(shape:[1, 4], scalars: [yOne, xOne, yTwo, xTwo])
+    let boxIndicies = Tensor<Int32>([0])
+
+    let centerCroppedImage = Raw.cropAndResize(image: Tensor<UInt8>([loadedJpeg]), boxes: boxesWrapped, boxInd: boxIndicies, cropSize:Tensor<Int32>([Int32(size.0),Int32(size.1)]))
```
Comment (rxwei) [inline Imagenette/data.swift]: Please fit all lines within 100 columns.

Comment (brettkoonce) [reply Imagenette/data.swift]: tweaked!

File: Imagenette/data.swift (hunk 5 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: removed!

File: Imagenette/main.swift (hunk 4 above)
Comment (brettkoonce) [reply Imagenette/main.swift]: done!

**Comment by brettkoonce** [comment]
@rxwei happy to wait, always willing to refactor!

File: Imagenette/data.swift [hunk 7]
```diff
@@ -38,7 +38,7 @@ struct ImageDataset {
         self.classes = newLabels.count
         self.labels = newLabels
         self.imageData = Tensor<Float>(
-            shape:[Int32(newImageLabels.count), Int32(imageSize.0), Int32(imageSize.1), 3],
+            shape:[Int(newImageLabels.count), Int(imageSize.0), Int(imageSize.1), 3],
```
Comment (rxwei) [inline Imagenette/data.swift]: You should be able to remove these conversions. They are already `Int`s.

Comment (brettkoonce) [reply Imagenette/data.swift]: I knew you were gonna say that! 😂 Let me look at it tomorrow, enough code for one day!

Comment (rxwei) [reply Imagenette/data.swift]: Cool!

File: Imagenette/data.swift [hunk 8]
```diff
@@ -38,35 +38,34 @@ struct ImageDataset {
...
     let loadedJpeg = Raw.decodeJpeg(contents: loadedFile, channels: 3, dctMethod: "")
 
-    let maxX: Float = Float(loadedJpeg.shape[0])
-    let maxY: Float = Float(loadedJpeg.shape[1])
+    let maxX = Float(loadedJpeg.shape[0])
+    let maxY = Float(loadedJpeg.shape[1])
 
-    let xPrime: Float = (Float(maxX) - Float(size.0))/2.0
-    let yPrime: Float = (Float(maxY) - Float(size.1))/2.0
+    let xPrime = (maxX - Float(size.0)) / 2.0
```
Comment (brettkoonce) [inline Imagenette/data.swift]: @rxwei conceptually, should I be reversing these params?  e.g. y = .0 and x = .1?  the math is the same when the width/height are equal.  this runs (slowly) with the latest cpu nightly.

File: Imagenette/data.swift (hunk 7 above)
Comment (brettkoonce) [reply Imagenette/data.swift]: see cleanup/question below!
